### PR TITLE
config: support determine terminal filter at runtime

### LINKED ISSUE #15565: Question: Is there are any way to determine whether is a terminal filter at runtime?
*Title*: *Is there are any way to determine whether is a terminal filter at runtime*

*Description*:
Put `is_terminal` flag at proto file, and return deferent filter logic base on the flag.

### PR DESCRIPTION
**Pull request by qinggniq**
fix #15565.

Currently we validate terminal filter before getting real protobuf message, now envoy can determine a filter whether is a terminal filter base on the protobuf message.

Commit Message: config: support determine terminal filter at runtime
Additional Description:
Risk Level: Low
Testing: TBD
Docs Changes: N/A
Release Notes: N/A
Platform Specific Features: N/A
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: include/envoy/filter/http/filter_config_provider.h [hunk 1]
```diff
@@ -32,11 +32,14 @@ class FilterConfigProviderManager {
    * @param filter_config_name the filter config resource name.
    * @param factory_context is the context to use for the filter config provider.
    * @param stat_prefix supplies the stat_prefix to use for the provider stats.
+   * @param last_filter_in_current_config indicate the filter position at filter chain
```
Comment (snowp) [inline include/envoy/filter/http/filter_config_provider.h:35]: maybe something like `indicates whether this filter is the last filter in the configured chain`

File: include/envoy/filter/http/filter_config_provider.h [hunk 2]
```diff
@@ -32,11 +32,14 @@ class FilterConfigProviderManager {
...
+   * @param filter_chain_type is the filter chain type
    */
   virtual FilterConfigProviderPtr createDynamicFilterConfigProvider(
       const envoy::config::core::v3::ExtensionConfigSource& config_source,
       const std::string& filter_config_name, Server::Configuration::FactoryContext& factory_context,
-      const std::string& stat_prefix) PURE;
+      const std::string& stat_prefix, bool last_filter_in_current_config,
```
Comment (snowp) [inline include/envoy/filter/http/filter_config_provider.h:41]: maybe just `last_filter`? or `last_filter_in_filter_chain`? not sure what "current" is referring to

File: source/common/filter/http/filter_config_discovery_impl.cc [hunk 3]
```diff
@@ -64,6 +69,12 @@ void DynamicFilterConfigProviderImpl::onConfigUpdate(Envoy::Http::FilterFactoryC
         this->current_config_ = config;
       });
 }
+void DynamicFilterConfigProviderImpl::validateTermianlFilter(const std::string& name,
```
Comment (snowp) [inline source/common/filter/http/filter_config_discovery_impl.cc:72]: typo in name

File: source/common/filter/http/filter_config_discovery_impl.cc [hunk 4]
```diff
@@ -106,7 +117,7 @@ void FilterConfigSubscription::onConfigUpdate(
     throw EnvoyException(fmt::format(
         "Unexpected number of resources in ExtensionConfigDS response: {}", resources.size()));
   }
-  const auto& filter_config = dynamic_cast<const envoy::config::core::v3::TypedExtensionConfig&>(
+  auto& filter_config = dynamic_cast<const envoy::config::core::v3::TypedExtensionConfig&>(
```
Comment (snowp) [inline source/common/filter/http/filter_config_discovery_impl.cc:120]: why remove the const here? we cast to a const ref

File: source/common/filter/http/filter_config_discovery_impl.h [hunk 5]
```diff
@@ -121,6 +129,7 @@ class FilterConfigSubscription
   absl::optional<Envoy::Http::FilterFactoryCb> last_config_{absl::nullopt};
   std::string last_type_url_;
   std::string last_version_info_;
+  envoy::config::core::v3::TypedExtensionConfig last_filter_config_;
```
Comment (snowp) [inline source/common/filter/http/filter_config_discovery_impl.h:132]: I think we intentionally avoid storing the entire filter config, can we just track whether the last filter passed the terminal filter check?

**Review by snowp — CHANGES_REQUESTED** [review]
Thanks this makes sense, just some questions around how we'll handle the dynamic case

File: source/common/filter/http/filter_config_discovery_impl.cc (hunk 4 above)
Comment (qinggniq) [reply source/common/filter/http/filter_config_discovery_impl.cc:120]: sorry, did not notice.

File: source/common/filter/http/filter_config_discovery_impl.h (hunk 5 above)
Comment (qinggniq) [reply source/common/filter/http/filter_config_discovery_impl.h:132]: Yes, we only need `it_terminal_filter`

File: include/envoy/filter/http/filter_config_provider.h (hunk 2 above)
Comment (qinggniq) [reply include/envoy/filter/http/filter_config_provider.h:41]: `last_filter_in_filter_chain` better.

File: include/envoy/filter/http/filter_config_provider.h (hunk 1 above)
Comment (qinggniq) [reply include/envoy/filter/http/filter_config_provider.h:35]: fixed.

File: source/common/filter/http/filter_config_discovery_impl.cc (hunk 3 above)
Comment (qinggniq) [reply source/common/filter/http/filter_config_discovery_impl.cc:72]: fixed.

**Comment by snowp** [comment]
Can you merge main to resolve conflicts and look at getting CI green? Thanks

**Comment by qinggniq** [comment]
@snowp I think the remain ci flakes is due to not update https://github.com/envoyproxy/envoy-filter-example, should I give a pr to that repo?

File: test/common/filter/http/filter_config_discovery_impl_test.cc [hunk 6]
```diff
@@ -353,7 +355,6 @@ TEST_F(FilterConfigDiscoveryImplTest, DualProviders) {
 TEST_F(FilterConfigDiscoveryImplTest, DualProvidersInvalid) {
   InSequence s;
   setup();
-  auto provider2 = createProvider("foo", true, false);
```
Comment (snowp) [inline test/common/filter/http/filter_config_discovery_impl_test.cc:356]: is this intentional?

File: test/integration/extension_discovery_integration_test.cc [hunk 7]
```diff
@@ -600,6 +608,28 @@ TEST_P(ExtensionDiscoveryIntegrationTest, BasicFailWithoutDefault) {
   EXPECT_EQ("500", response->headers().getStatusValue());
 }
 
+// test when terminal filter config not put at last config position
```
Comment (snowp) [inline test/integration/extension_discovery_integration_test.cc:611]: nit: proper sentence, proper casing + punctuation

File: test/integration/filters/set_response_code_filter_config.proto [hunk 8]
```diff
@@ -8,4 +8,5 @@ message SetResponseCodeFilterConfig {
   string prefix = 1;
   uint32 code = 2 [(validate.rules).uint32 = {lt: 600 gte: 200}];
   string body = 3;
+  bool is_terminal_filter = 4;
```
Comment (snowp) [inline test/integration/filters/set_response_code_filter_config.proto:11]: I would do this in a different test filter instead of overloading the existing one

**Review by snowp — CHANGES_REQUESTED** [review]
Thanks, just a couple comments

CI seems to still be broken

/wait

File: test/common/filter/http/filter_config_discovery_impl_test.cc (hunk 6 above)
Comment (qinggniq) [reply test/common/filter/http/filter_config_discovery_impl_test.cc]: By mistake.

File: test/integration/extension_discovery_integration_test.cc (hunk 7 above)
Comment (qinggniq) [reply test/integration/extension_discovery_integration_test.cc]: fixed.

File: test/integration/filters/set_response_code_filter_config.proto (hunk 8 above)
Comment (qinggniq) [reply test/integration/filters/set_response_code_filter_config.proto]: reset.

**Comment by qinggniq** [comment]
@snowp , the ci seems failed due to the third-party repo is not updated, as I commented at https://github.com/envoyproxy/envoy/issues/15565#issuecomment-806323001 . It seems to become a cycle-dependent:
- If I give a pr for  https://github.com/envoyproxy/envoy-filter-example , it will fail to build due to not get this commit.
- If I give a pr to envoy like this pr, it failed due to not update `echo filter` in envoy-filter-example.

File: test/extensions/filters/network/http_connection_manager/config_test.cc [hunk 9]
```diff
@@ -1850,6 +1850,7 @@ stat_prefix: router
     config_source: { resource_api_version: V3, ads: {} }
     default_config:
       "@type": type.googleapis.com/envoy.extensions.filters.http.health_check.v3.HealthCheck
+      pass_through_mode: false
```
Comment (snowp) [inline test/extensions/filters/network/http_connection_manager/config_test.cc:1853]: Why this change?

File: test/integration/extension_discovery_integration_test.cc [hunk 10]
```diff
@@ -692,5 +707,28 @@ TEST_P(ExtensionDiscoveryIntegrationTest, DestroyDuringInit) {
   ecds_connection_.reset();
 }
 
+// Validate that a listener update should fail if the subscribed extension configuration make filter
+// terminal but the filter position is not at the last position at filter chain.
+TEST_P(ExtensionDiscoveryIntegrationTest, BasicFailTerminalFilterNotFit) {
```
Comment (snowp) [inline test/integration/extension_discovery_integration_test.cc:712]: The `NotFit` bit here is a bit vague, can you make it clearer from the name what we're testing here?

File: test/server/listener_manager_impl_test.cc [hunk 11]
```diff
@@ -486,26 +486,6 @@ TEST_F(ListenerManagerImplWithRealFiltersTest, TerminalNotLast) {
       "filter in a network filter chain.");
 }
 
-TEST_F(ListenerManagerImplWithRealFiltersTest, NotTerminalLast) {
-  const std::string yaml = R"EOF(
```
Comment (snowp) [inline test/server/listener_manager_impl_test.cc:490]: Can you revert these changes if they make no functional change? Just to better preserve git history

**Review by snowp — CHANGES_REQUESTED** [review]
Thanks just a few small comments. WRT the examples repo I think we have to break that with this change and then fix it up later on, I don't think there's a way for us to get this fully green

@alyssawilk do you want to take a look as well since this touches the terminal filter stuff you added?

**Comment by alyssawilk** [comment]
@snowp I'm happy to let you make the call of what makes sense, but if you want me to do a second pass I'm game!

**Comment by snowp** [comment]
All good then, I'll handle the review :)

File: test/extensions/filters/network/http_connection_manager/config_test.cc (hunk 9 above)
Comment (qinggniq) [reply test/extensions/filters/network/http_connection_manager/config_test.cc:1853]: ref the HealthCheck proto:
https://github.com/envoyproxy/envoy/blob/5b1977f607962d7b700ce486651ca12bd5659851/api/envoy/extensions/filters/http/health_check/v3/health_check.proto#L31-L32, it requires `pass_through_mode` to be set. Otherwise, it will raise validate error but not terminal filter not fit error.

Comment (snowp) [reply test/extensions/filters/network/http_connection_manager/config_test.cc:1853]: For that to be the case won't the HealthCheck extension's isTerminalFilter function need to be updated in this PR to read the config? I don't see a change like that, but maybe I'm missing something?

File: source/extensions/filters/network/http_connection_manager/config.cc [hunk 12]
```diff
@@ -524,7 +524,7 @@ void HttpConnectionManagerConfig::processFilter(
       proto_config, context_.messageValidationVisitor(), factory);
   Http::FilterFactoryCb callback =
       factory.createFilterFactoryFromProto(*message, stats_prefix_, context_);
-  bool is_terminal = factory.isTerminalFilter();
+  bool is_terminal = factory.isTerminalFilterByProto(*message, context_);
```
Comment (qinggniq) [inline source/extensions/filters/network/http_connection_manager/config.cc:527]: Here we should read the message before validating the filter whether it's a terminal filter.

File: test/extensions/filters/network/http_connection_manager/config_test.cc (hunk 9 above)
Comment (qinggniq) [reply test/extensions/filters/network/http_connection_manager/config_test.cc:1853]: Most HTTP filter class is inheriting from `Common::FactoryBase<T>`, and use the default isTerminalFilter function from the parent, I changed it at this class.

Comment (snowp) [reply test/extensions/filters/network/http_connection_manager/config_test.cc:1853]: I don't see you changing the the function for the HealthCheck filter? Shouldn't it be using the default FactoryBase one that always returns false?

It seems right to me that the HealthCheck filter should change isTerminal based on this flag, I just don't see that code included in this PR

Comment (qinggniq) [reply test/extensions/filters/network/http_connection_manager/config_test.cc:1853]: 1. I changed the `isTerminalFilter` function at the base class `Common::FactoryBase<T>`, which is the base class of HealthCheck filter factory. https://github.com/qinggniq/envoy/blob/b9dfe5c0a780ca8d08230aa9aeace6e3e6cf4c47/source/extensions/filters/http/common/factory_base.h#L45-L50
2. I change the validating sequence from `terminal filter validate -> parse protocol message (validate message) -> create filter` to `parse protocol message (validate message)  -> terminal filter validate  -> create filter` because now isTerminalFilter rely on message.
3. I added `pass_through_mode = 1` because after point 2 I mentioned, the previous validate sequence will not parse the message because it will raise *terminal filter to validate error* first, after change the validating sequence,  it will raise `validate message error` first, which it's not we want to test.

Do these clarify clear?

Comment (snowp) [reply test/extensions/filters/network/http_connection_manager/config_test.cc:1853]: Oh I see, this is about message validation, not about it affecting terminal filters. Thanks for explaining!

File: test/integration/extension_discovery_integration_test.cc (hunk 10 above)
Comment (qinggniq) [reply test/integration/extension_discovery_integration_test.cc:712]: How about BasicFailTerminalFilterNotAtLastInFilterChain?

File: test/server/listener_manager_impl_test.cc (hunk 11 above)
Comment (qinggniq) [reply test/server/listener_manager_impl_test.cc:490]: fixed.

File: test/integration/extension_discovery_integration_test.cc (hunk 10 above)
Comment (snowp) [reply test/integration/extension_discovery_integration_test.cc:712]: `BasicFailTerminalFilterNotAtEndOfFilterChain` maybe?

Comment (qinggniq) [reply test/integration/extension_discovery_integration_test.cc:712]: ok, fixed.

**Review by snowp — DISMISSED** [review]
Thanks this LGTM. Can you check CI? I think the asan/tsan ones are legit

**Comment by snowp** [comment]
Oh actually no those are the filter example ones you mentioned, let me look into how we can fix this

**Comment by snowp** [comment]
@qinggniq Can you make a PR to envoy-filter-example and point the SHA in ci/filter_example_setup.sh to the commit in the PR? That should let this PR test against the envoy-filter-example PR

**Comment by qinggniq** [comment]
The failures of ci `asan` and `tsan` seem due to https://github.com/envoyproxy/envoy/pull/15972#issuecomment-820214392.

**Comment by qinggniq** [comment]
@snowp all cis are green, waiting for https://github.com/envoyproxy/envoy-filter-example/pull/144 merged. I will change the commit SHA and URL at ci/setup_example_filter.sh to the latest commit.

**Comment by snowp** [comment]
I merged the filter example PR, can you update this to use the new commit and we'll get this landed? Thanks

**Review by snowp — APPROVED** [review]
Thanks!
